graphblas_algorithms/cluster.py

```diff
@@ -0,0 +1,73 @@
+from collections import OrderedDict
+
+import graphblas as gb
+from graphblas import Matrix, Vector, binary, select
+from graphblas.semiring import any_pair, plus_pair
+from networkx.utils import not_implemented_for
+
+
+def single_triangle_core(G, index, *, L=None, has_self_edges=True):
+    M = Matrix(bool, G.nrows, G.ncols)
+    M[index, index] = False
+    C = any_pair(G.T @ M.T).new(name="C")  # select.coleq(G.T, index)
+    del C[index, index]  # Ignore self-edges
+    R = C.T.new(name="R")
+    if has_self_edges:
+        if L is None:
+            # Pretty much all the time is spent here.
+            # We take the TRIL as a way to ignore the self-edges.
+            L = select.tril(G, -1).new(name="L")
+        return plus_pair(L @ R.T).new(mask=C.S).reduce_scalar(allow_empty=False).value
+    else:
+        return plus_pair(G @ R.T).new(mask=C.S).reduce_scalar(allow_empty=False).value // 2
+
+
+def triangles_core(G, mask=None, *, L=None, U=None):
+    # Ignores self-edges
+    if L is None:
+        L = select.tril(G, -1).new(name="L")
+    if U is None:
+        U = select.triu(G, 1).new(name="U")
+    C = plus_pair(L @ L.T).new(mask=L.S)
+    return (
+        C.reduce_rowwise().new(mask=mask)
+        + C.reduce_columnwise().new(mask=mask)
+        + plus_pair(U @ L.T).new(mask=U.S).reduce_rowwise().new(mask=mask)
+    ).new(name="triangles")
```

Comment:
I wish I had an intuitive explanation for why this worked (I'm sure one exists if you twist your brain *just so*).  There are simpler methods, but this implementation seems pretty fast.

Here's an example of a simpler method:
```python
    L = gb.select.tril(G, -1).new(name="L")
    B = gb.select.offdiag(G).new(name="offdiag")
    return plus_pair(B @ L.T).new(mask=B.S).reduce_rowwise().new(mask=mask)
```